### PR TITLE
Validate URLs before saving as metadata

### PR DESCRIPTION
**Pull request by eunices**
**What this PR does / why we need it**:
Validate URLs before saving as metadata.

**Which issue(s) this PR closes**:
NA

**Special notes for your reviewer**:
None

**Suggestions on how to test this**:
Add `http://google.com test` to Alternative URL when editing metadata. It should not work.
Only https://, http:// and ftp:// allowed.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
No

**Is there a release notes update needed for this change?**:
No

**Additional documentation**:
None

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java [hunk 1]
```diff
@@ -168,7 +169,20 @@ public boolean isValid(DatasetFieldValue value, ConstraintValidatorContext conte
 
         if (fieldType.equals(FieldType.URL) && !lengthOnly) {
             try {
-                URL url = new URL(value.getValue());
+                
+                String[] schemes = {"http","https"};
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java:173]: Other than being less popular, is there any reason ftp URLs shouldn't be allowed?

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java [hunk 2]
```diff
@@ -168,7 +169,20 @@ public boolean isValid(DatasetFieldValue value, ConstraintValidatorContext conte
...
+                UrlValidator urlValidator = new UrlValidator(schemes);
+                
+                String urlString = value.getValue();
+                URL url = new URL(urlString);
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java:177]: FWIW: Not sure if this and the catch of a MalformedURLException is still needed. The urlValidator class tries to create a java URI and returns false if that gives a MalformedURLException

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 import javax.validation.ConstraintValidator;
 import javax.validation.ConstraintValidatorContext;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.validator.routines.UrlValidator;
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java:22]: There's a similarly named edu.harvard.iq.dataverse.URLValidator class that looks like it was used in bean validation (pre v4.13) and in a couple places to validate URLs from settings. I don't see any place where it it used for displayed URLs. Never-the-less, seems like using the same Apache UrlValidator class there makes sense.

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java (hunk 2 above)
Comment (eunices) [reply src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java]: I see, removed MalformedURLException.

Comment (eunices) [reply src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java]: Indeed, removed `URL url = new URL(urlString)`.

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java (hunk 1 above)
Comment (eunices) [reply src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java]: Nope, have added ftp in.

File: src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java (hunk 3 above)
Comment (eunices) [reply src/main/java/edu/harvard/iq/dataverse/DatasetFieldValueValidator.java:19]: Ok, have updated URLValidator to use org.apache.commons.validator.routines.UrlValidator, see 6de1e959555c315f79826e8ab86796a696aa49ee. It passed the corresponding unit test.

**Comment by eunices** [comment]
I've addressed your comments @qqmyers sufficiently, I hope!

**Review by sekmiller — APPROVED** [review]
Looks good. And I checked in with Jim regarding his previous comments. He agrees that it's ready to move.

**Pull request merged**